src/Data/Fuzzy/Fuzzy.purs

```diff
@@ -0,0 +1,222 @@
+module Data.Fuzzy where
+
+import Prelude
+
+import Data.Array (snoc, unsnoc)
+import Data.Either (Either(..))
+import Data.Foldable (all, foldl)
+import Data.Generic.Rep (class Generic)
+import Data.Generic.Rep.Eq (genericEq)
+import Data.Generic.Rep.Ord (genericCompare)
+import Data.Generic.Rep.Show (genericShow)
+import Data.StrMap (StrMap, values)
+import Data.Maybe (Maybe(..), fromMaybe)
+import Data.Monoid (mempty)
+import Data.Newtype (class Newtype, unwrap)
+import Data.String (Pattern(..), drop, indexOf, lastIndexOf, take, toLower)
+import Data.String.Utils (length, toCharArray, words)
+import Data.Tuple (Tuple(..))
+
+newtype Fuzzy a = Fuzzy
+  { original :: a
+  , result   :: StrMap (Maybe Result)
+  , score    :: Rank
+  }
+
+derive instance genericFuzzy :: Generic (Fuzzy a) _
+derive instance newtypeFuzzy :: Newtype (Fuzzy a) _
+instance eqFuzzy :: Eq a => Eq (Fuzzy a) where eq = genericEq
+instance showFuzzy :: Show a => Show (Fuzzy a) where show = genericShow
+instance ordFuzzy :: Eq a => Ord (Fuzzy a) where
+  compare (Fuzzy { score }) (Fuzzy { score: score' }) = compare score score'
+
+newtype FuzzyStr = FuzzyStr
+  { result :: Result
+  , score  :: Rank
+  }
+
+derive instance genericFuzzyStr :: Generic FuzzyStr _
+derive instance newtypeFuzzyStr :: Newtype FuzzyStr _
+instance eqFuzzyStr :: Eq FuzzyStr where eq = genericEq
+instance showFuzzyStr :: Show FuzzyStr where show = genericShow
+instance ordFuzzyStr :: Ord FuzzyStr where
+  compare (FuzzyStr { score }) (FuzzyStr { score: score' }) = compare score score'
+
+data Depth = Full | Word | Char
+
+derive instance genericDepth :: Generic Depth _
+instance eqDepth :: Eq Depth where eq = genericEq
+instance showDepth :: Show Depth where show = genericShow
+
+data Pos = Start | Prefix | Mid | Suffix | End
+
+derive instance genericPos :: Generic Pos _
+instance eqPos :: Eq Pos where eq = genericEq
+instance showPos :: Show Pos where show = genericShow
+
+data Rank = Rank Int Int Int Int Int Int | None
```

Comment:
I know we've discussed this in person, but I could use a comment here briefly noting how the ranking works (0 0 0 0 0 0 is an exact match, higher numbers are worse, `None` indicates nothing matched at all). Something quick and dirty is fine.

Reply:
Yeah this was a premature PR, I do intend to go through and add thorough documentation, should've done so beforehand